ddht/v5_1/alexandria/partials/chunking.py

```diff
@@ -0,0 +1,221 @@
+    """
+    Given a chunk index, convert it to the path into the binary tree where the
+    chunk is located.
+    """
+    for power_of_two in reversed(POWERS_OF_TWO[:path_bit_size]):
+        yield bool(index & power_of_two)
+
+
+def path_to_left_chunk_index(path: TreePath, path_bit_size: int) -> int:
+    """
+    Given a path, convert it to a chunk index.  In the case where the path is
+    to an intermediate tree node, return the chunk index on the leftmost branch
+    of the subtree.
+    """
+    return sum(
+        power_of_two
+        for path_bit, power_of_two in itertools.zip_longest(
+            path, reversed(POWERS_OF_TWO[:path_bit_size]), fillvalue=False,
+        )
+        if path_bit
+    )
+
+
+@to_tuple
+def group_by_subtree(
+    first_chunk_index: int, num_chunks: int
+) -> Iterable[Tuple[int, ...]]:
+    r"""
+    Group the paths into groups that belong to the same subtree.
+
+    This helper function is used when constructing partial proofs. After
+    setting aside the leaf nodes that correspond to the data we wish to prove,
+    the remaining leaf nodes can be replaced by the hashes of the largest
+    subtrees that contains them.
+
+    Given a 4-bit tree like this
+
+    0:                           0
+                                / \
+                              /     \
+                            /         \
+                          /             \
+                        /                 \
+                      /                     \
+                    /                         \
+    1:             0                           1
+                 /   \                       /   \
+               /       \                   /       \
+             /           \               /           \
+    2:      0             1             0             1
+          /   \         /   \         /   \         /   \
+    3:   0     1       0     1       0     1       0     1
+        / \   / \     / \   / \     / \   / \     / \   / \
+    4: A   B C   D   E   F G   H   I   J K   L   M   N O   P
+
+                |-------------------------|
+
+    If we are given the chunks D-K which map to indices 3-10 we want them
+    divided up into the largest subgroups that all belong in the same subtree.
+
+    2:      0             1             0             1
+          /   \         /   \         /   \         /   \
+    3:   0     1       0     1       0     1       0     1
+        / \   / \     / \   / \     / \   / \     / \   / \
+    4: A   B C   D   E   F G   H   I   J K   L   M   N O   P
+                |-| |-----------| |-----|-|
+
+    These groups are
+
+    - D
+    - E, F, G, H
+    - I, J
+    - K
+
+    The algorithm for doing this is as follows:
+
+    1) Find the largest power of 2 that can fit into the chunk range.  This is
+    referred to as the `group_size`.  For this example the number is 8 since we
+    have a span of 8 items.  Divide the range up into groups aligned with this
```

Comment:
This walkthrough is excellent. Still took me a couple reads to grok though. Couple minor nits/notes: 
- first time(s) through, I stumbled over differentiating `groups`, `ranges`, & `chunks`. I found the bucket metaphor helpful, and that may be an improvement over one of these terms.
- can simplify to ~"Divide the range up into into groups of this number."
- typo: `lenght`
- "Any group that is the full length __*and within the bucket*__ is final".
- (These type of onboarding/documentation improvements can def be prioritized later)